### PR TITLE
context-aware electron support 

### PR DESCRIPTION
**Pull request by d4tocchini**
FYI, I stumbled my way to the #v1-test branch using @DoctorEvidence's excellent lmdb-store for an electron project. Starting with latest electron (v9), renderer process reuse is on by default, as it should be given the perf implications. This makes context-aware a must, [see](https://nodejs.org/dist/latest-v12.x/docs/api/addons.html#addons_context_aware_addons).

Do you expect your branch to be merged with @Venemo's master?

### CODE REVIEW

**Comment by kriszyp** [comment]
Does this make node-lmdb thread-safe? I'd be glad to give this a try, that sounds like it would be valuable improvement.

> Do you expect your branch to be merged with @Venemo's master?

I do. I have accumulated a bit of fixes/improvements that I needed for lmdb-store and our projects that use it, and needed to be able to use them right away, hence the tag referencing. But generally, Venemo does usually merge my PRs, it usually just takes a while since he is not very actively working on the project.

**Comment by d4tocchini** [comment]
I wouldn't say it's thread-safe, or node worker_thread-safe, without proper passing tests, but I can confirm it is working with electron renderer process reuse.  

The last remaining issue I see arising with worker_threads has to do with the following from [LMDB: Getting Started #Threads and Processes](http://www.lmdb.tech/doc/starting.html):

> LMDB uses POSIX locks on files, and these locks have issues if one process opens a file multiple times. Because of this, do not `mdb_env_open()` a file multiple times from a single process. Instead, share the LMDB environment that has opened the file across all threads. Otherwise, if a single process opens the same environment multiple times, closing it once will remove all the locks held on it, and the other instances will be vulnerable to corruption from other processes.

Thus, issues may arise when multiple worker_threads `mdb_env_open` or `mdb_env_close` the same file.  Currently, this PR should work as long as each worker_thread env_opens/closes different files.  But, of course, that's not good enough :)  I plan on tackling this soon with specs...  Something like wrapping env_open/close usage with logic that checks a thread-shared data structure to track how many threads have an env file opened and shares the env handle.  When opening an env, if thread count > 0 then return a shared env handle. When closing an env, if thread count goes to zero then & only then `mdb_env_close`.  Make sense?  Let me know if you any ideas / preferences on the approach.

**Comment by d4tocchini** [comment]
A further optimization would be handling batch writes cross-worker_threads.

A counter-point could be made to forgo allowing worker_threads to open the same env in favor of taking advantage of sharedarraybuffers in JSland...  Hmmm...

**Comment by d4tocchini** [comment]
Ha, I just realized @DoctorEvidence is an org, not a user, and you are the brains behind those repos.  Hats-off to you, sir

File: src/env.cpp [hunk 1]
```diff
@@ -163,17 +163,18 @@ struct action_t {
     argtokey_callback_t freeKey;
 };
 
-int deleteValue; // pointer to this as the value represents a delete
+thread_local static int deleteValue; // pointer to this as the value represents a delete
```
Comment (kriszyp) [inline src/env.cpp:166]: I wouldn't think we would actually need to change deleteValue for threads, it is merely being used a static pointer location (to differentiate between NULL), there is never any value assigned or changed. There is probably a much more correct way of doing this (having a reserved constant pointer that won't conflict with other pointers). All the other static variables do need to change though.

**Comment by kriszyp** [comment]
Yes, that makes sense.

FWIW, I think the general use case for using threads with LMDB in Node is somewhat weak. Node's threads are kind of heavy, and the main reason for threading is shared memory. And LMDB essentially actually provides shared memory for processes (with memory-maps, processes are really sharing the same memory), in a far more robust way than any SharedArrayBuffer, it is essentially transactional memory, giving you the benefit of threads with multiple processes.

That being said, there are plenty of use cases for threads in Node, and being able to access LMDB in them would be a nice addition, so I applaud your efforts.

And I can certainly merge your current work into my fork, so that lmdb-store will work in electron, if that is your immediate need.

**Comment by kriszyp** [comment]
Also, are you going to submit this as a PR to the main repo? It would be nice to have this directly merged upstream (and you have already commented on the issue in the main repo).

File: src/env.cpp (hunk 1 above)
Comment (d4tocchini) [reply src/env.cpp:166]: if it's read-only, makes sense

**Pull request merged**